### ISSUE TITLE
What's the recommended approach to install (recent) open-vm-tools on Ubuntu 16.04 Desktop?

### ISSUE THREAD

**Issue by smattheis** [vmware/open-vm-tools#229]
I'm running a VM with Ubuntu 16.04 LTS Desktop and recently upgraded from VMWare Fusion 8 to VMware Fusion 10. Since VMWare Fusion 10 recommends (!) to use open-vm-tools instead of vmware-tools, I uninstalled vmware-tools and installed open-vm-tools from aptitude and restarted:
``` bash
sudo apt-get install open-vm-tools open-vm-tools-desktop open-vm-tools-dkms
```
In that setup, I have a shared folder for source code repositories maintained with git. However, this setup breaks because of some bug in vmhgfs-fuse  as reported in https://github.com/vmware/open-vm-tools/issues/90 which, e.g., causes a git rebase to loose commits (!!!). (That's a nightmare!) I know that the bug was fixed with version 10.1.0, but standard package repositories for ubuntu 16.04 provide only version 10.0.7:
``` bash
$ apt-show-versions open-vm-tools
open-vm-tools:amd64/xenial-updates 2:10.0.7-3227872-5ubuntu1~16.04.1 uptodate
open-vm-tools:i386 not installed
``` 
Since I couldn't find any official and reliable PPA for more recent packages of open-vm-tools, see https://launchpad.net/ubuntu/+ppas?name_filter=open-vm-tools, I tried the "build it yourself" way:
``` bash
git clone https://github.com/vmware/open-vm-tools.git
cd open-vm-tools
git checkout stable-10.2.0 -b stable-10.2.0
cd open-vm-tools
sudo apt-get install libmspack-dev
./configure
...
checking for libmspack >= 0.0.20040308alpha (via pkg-config)... yes
checking for glib-2.0 >= 2.14.0 (via pkg-config)... no
configure: error: glib2 >= 2.14.0 is required.
```
At this point, I'm lost: I have no idea how to install glib2 on ubuntu 16.04 nor if that's even possible nor do I know if version 10.2.0 is suitable for ubuntu 16.04.

In short:
1. Is there a (official/reliable) PPA for open-vm-tools for more recent versions of open-vm-tools?
2. If not, what's the recommended way to install recent versions of open-vm-tools such as 10.1.0+?
3. If it's "build it yourself", how do I build open-vm-tools for Ubuntu 16.04 **Desktop**? (Should I use a specific version? Is there a recommended set of configure options?)

(Please understand that it's not an option for me to upgrade Ubuntu now because the next LTS is 18.04 but experience shows that upgrading to some Ubuntu LTS xy.04 is not a good idea before xy.10.)

**Comment by TheHellSite** [comment]
Very interested in this one also!

**Comment by osde8info** [comment]
linuxlite (based on 16.04) works fine with vmware player 14 and open-vm-tools
maybe you could try that in a new fusion vm

ps fusion 10 wont run at all on my 2010 macbook so i have to stick with fusion 8

**Comment by osde8info** [comment]
just tried to build open-vm-tools on linuxlite 3.8 and i also get
`
checking for libintl.h... yes
checking for libmspack >= 0.0.20040308alpha (via pkg-config)... no
configure: error: mspack >= 0.0.20040308alpha is required.
`

**Comment by osde8info** [comment]
which is fixed by

`
# aptitude install libmspack-dev
The following NEW packages will be installed:
  libmspack-dev 
0 packages upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 54.5 kB of archives. After unpacking 213 kB will be used.
Get: 1 http://archive.ubuntu.com/ubuntu xenial-security/main amd64 libmspack-dev amd64 0.5-1ubuntu0.16.04.1 [54.5 kB]
Fetched 54.5 kB in 0s (125 kB/s)   

Selecting previously unselected package libmspack-dev:amd64.
(Reading database ... 203622 files and directories currently installed.)
Preparing to unpack .../libmspack-dev_0.5-1ubuntu0.16.04.1_amd64.deb ...
Unpacking libmspack-dev:amd64 (0.5-1ubuntu0.16.04.1) ...
Setting up libmspack-dev:amd64 (0.5-1ubuntu0.16.04.1) ...
`

**Comment by osde8info** [comment]
libglib2 erroror seems to be fixed by
`# aptitude install libglib2.0-dev`

**Comment by osde8info** [comment]
ok now i get
```
checking for fuse (via pkg-config)... no
configure: WARNING: Fuse is missing, vmblock-fuse/vmhgfs-fuse will be disabled.
checking security/pam_appl.h usability... no
checking security/pam_appl.h presence... no
checking for security/pam_appl.h... no
configure: error: Cannot find PAM library. Please configure without PAM (using --without-pam), or install the PAM libraries and devel package(s).
```

**Comment by oliverkurth** [comment]
Here is a quick guide:

Add this line to `/etc/apt/sources.list`: 
`deb-src http://us.archive.ubuntu.com/ubuntu/ xenial main restricted
`
Then do:
```
apt-get update
apt-get install -y build-essential
apt-get build-dep -y open-vm-tools
apt-get install -y libmspack-dev libssl-dev libxerces-c-dev libxml-security-c-dev libprocps4-dev

```
See https://github.com/vmware/open-vm-tools/issues/98#issuecomment-284032194

However, building open-vm-tools this way won't give you a functioning install. I am currently testing a better solution.

**Comment by osde8info** [comment]
thanks @oliverkurth it fixed lots of errors but i still get

# aptitude build-dep open-vm-tools
Unable to satisfy the build-depends: Build-Depends: libprocps-dev.
Unable to satisfy the build-depends: Build-Depends: libprocps-dev.
The following NEW packages will be installed:
  dh-systemd doxygen gir1.2-gtk-2.0{a} libatk1.0-dev{a} libatkmm-1.6-dev{a} libcairo-script-interpreter2{a} 
  libcairo2-dev{a} libcairomm-1.0-dev{a} libclang1-3.6{a} libcunit1{a} libcunit1-dev libdumbnet-dev 
  libdumbnet1{a} libfontconfig1-dev{a} libfreetype6-dev{a} libgdk-pixbuf2.0-dev{a} libglibmm-2.4-dev{a} 
  libgtk2.0-dev libgtkmm-2.4-dev libharfbuzz-dev{a} libharfbuzz-gobject0{a} libllvm3.6v5{a} libnotify-dev 
  libobjc-5-dev{ab} libobjc4{ab} libpango1.0-dev{a} libpangomm-1.4-dev{a} libpixman-1-dev{a} libpng12-dev{a} 
  libsigc++-2.0-dev{a} libxcb-render0-dev{a} libxcb-shm0-dev{a} libxcomposite-dev{a} libxcursor-dev{a} 
  libxdamage-dev{a} libxfixes-dev{a} libxft-dev{a} libxi-dev{a} libxinerama-dev libxml2-utils{a} 
  libxrandr-dev{a} libxrender-dev{a} libxss-dev libxtst-dev x11proto-composite-dev{a} x11proto-damage-dev{a} 
  x11proto-fixes-dev{a} x11proto-randr-dev{a} x11proto-record-dev{a} x11proto-render-dev{a} 
  x11proto-scrnsaver-dev{a} x11proto-xinerama-dev{a} 
0 packages upgraded, 52 newly installed, 0 to remove and 1 not upgraded.
Need to get 24.2 MB of archives. After unpacking 118 MB will be used.
The following packages have unmet dependencies:
 libobjc4 : Depends: gcc-5-base (= 5.4.0-6ubuntu1~16.04.6) but 5.4.0-6ubuntu1~16.04.9 is installed.
 libobjc-5-dev : Depends: gcc-5-base (= 5.4.0-6ubuntu1~16.04.6) but 5.4.0-6ubuntu1~16.04.9 is installed.
                 Depends: libgcc-5-dev (= 5.4.0-6ubuntu1~16.04.6) but 5.4.0-6ubuntu1~16.04.9 is installed.
The following actions will resolve these dependencies:

     Keep the following packages at their current version:
1)     doxygen [Not Installed]                            
2)     libclang1-3.6 [Not Installed]                      
3)     libobjc-5-dev [Not Installed]                      
4)     libobjc4 [Not Installed]

**Comment by oliverkurth** [comment]
@osde8info I don't see those errors when I try that myself. Can you post your file` /etc/apt/sources.list`?

**Comment by smattheis** [comment]
Thanks for the comments. That helped a bit. Now, I do the following preparation:
``` bash
$ apt-add-repository -s 'deb http://us.archive.ubuntu.com/ubuntu/ xenial main restricted'
$ apt-get update
$ apt-get install -y build-essential
$ apt-get build-dep -y open-vm-tools
Reading package lists... Done
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  autopoint debhelper dh-autoreconf dh-strip-nondeterminism dh-systemd doxygen gir1.2-gtk-2.0 icu-devtools libatk1.0-dev libatkmm-1.6-dev libcairo-script-interpreter2 libcairo2-dev libcairomm-1.0-dev libclang1-3.6 libcunit1 libcunit1-dev libdumbnet-dev libfile-stripnondeterminism-perl libfuse-dev libgdk-pixbuf2.0-dev libglib2.0-dev libglibmm-2.4-dev libgtk2.0-dev libgtkmm-2.4-dev
  libharfbuzz-dev libharfbuzz-gobject0 libicu-dev libllvm3.6v5 libnotify-dev libobjc-5-dev libobjc4 libpam0g-dev libpango1.0-dev libpangomm-1.4-dev libpcre3-dev libpcre32-3 libpcrecpp0v5 libpixman-1-dev libprocps4-dev libselinux1-dev libsepol1-dev libsigc++-2.0-dev libssl-dev libxcb-render0-dev libxcb-shm0-dev libxcomposite-dev libxcursor-dev libxdamage-dev libxft-dev
  libxinerama-dev libxrandr-dev libxss-dev po-debconf x11proto-composite-dev x11proto-damage-dev x11proto-randr-dev x11proto-scrnsaver-dev x11proto-xinerama-dev
0 upgraded, 58 newly installed, 0 to remove and 3 not upgraded.
Need to get 36.2 MB of archives.
After this operation, 176 MB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu xenial/main amd64 libdumbnet-dev amd64 1.12-7 [55.9 kB]
...
$ apt-get install -y libmspack-dev libssl-dev libxerces-c-dev libxml-security-c-dev libprocps4-dev
```
After that, I can successfully build from source:
``` bash
$ git clone https://github.com/vmware/open-vm-tools.git
$ cd open-vm-tools
$ git checkout stable-10.2.0 -b stable-10.2.0
$ cd open-vm-tools
$ autoreconf -i
$ ./configure --without-gtk3
$ make
$ ldconfig
```
The install binaries, I do the following using checkinstall (`sudo apt-get install checkinstall`):
``` bash
$ apt-get purge open-vm-tools
# Here, I needed to comment out any vmhgfs mounts from /etc/fstab because booting Ubuntu fails otherwise. Initially, I wasn't aware of that such that I had to do a little recovery via Ubuntu Live CD and chroot to finally comment it out.
$ shutdown -r now
$ sudo checkinstall --pkgname=open-vm-tools-from-source --pkgversion=10.2.0
...
$ shutdown -r now
```
The installation and package creation succeeds and binaries are installed under /usr/local/. However, none of the services work and I assume that's because the vmtoolsd is not started. Is that right? Should it be started? How do I do that? Just running `sudo vmtoolsd` doesn't do anything and the help option doesn't tell that much about it.

**Comment by oliverkurth** [comment]
Sorry @smattheis for not coming back to this sooner.

It's good that you were able to build open-vm-tools from source, but I mentioned earlier that this will not give you a functioning install, as you have noticed. The reason is that we do not ship any startup scripts (like SysV init scripts or systemd unit files), and the reason for this is that this is different for different Linux distros. So you would need to do that yourself, ideally adding system unit files. You can use the ones that ship with the Ubuntu package and change them.

There are other options though:
1) there *Is* a ppa for open-vm-tools here: https://launchpad.net/~ci-train-ppa-service/+archive/ubuntu/3152 . This is a pretty new, and may have issues though. I wasn't aware of this until I noted this issue on launchpad: https://bugs.launchpad.net/ubuntu/+source/open-vm-tools/+bug/1741390 . I recommend that you vote for it to be fixed.

2) less ideally, you can install VMware Tools. You can either use the iso image that ships with Fusion or Workstation by selecting "Install VMware Tools" in the menu, or download from here: https://my.vmware.com/web/vmware/details?downloadGroup=VMTOOLS1020&productId=614

**Comment by smattheis** [comment]
Thanks for the information. I wasn't aware of the issue and discussion going on on launchpad at the same time we were working on this problem. So if I got it right, it's actually kind of fortune that just few days after I created this issue a new PPA was added? Anyways, thanks for your help and thanks to @cpaelzer. For the records, the solution that finally worked for me is as follows:

``` bash
sudo apt-get purge open-vm-tools open-vm-tools-desktop open-vm-tools-dkms
# Here, I did a restart to be sure there is no conflict BUT I needed to COMMENT OUT vmhgfs-fuse mounts in /etc/fstab because otherwise reboot fails.
sudo shutdown -r now
sudo add-apt-repository ppa:ci-train-ppa-service/3152
sudo apt-get update
sudo apt-get install open-vm-tools open-vm-tools-desktop open-vm-tools-dkms
# Here, uncomment vmhgfs-fuse mounts in /etc/fstab.
sudo shutdown -r now
apt-show-versions open-vm-tools
open-vm-tools:amd64/xenial 2:10.2.0-3ubuntu0.16.04.1~ppa1 uptodate
open-vm-tools:i386 not installed
```

**Comment by oliverkurth** [comment]
Very nice! Please be aware that the ppa is experimental.

**Comment by cpaelzer** [comment]
Thanks @oliverkurth to point to the ppa.
As you already outlined it is experimental atm.

But @smattheis you can help to fix that.
What currently is needed are:
- most importantly people that tested the ppa and confirm it works (so that I can go on pushing it to -backports)
- more people to show demand for it on the bug never hurts

**Comment by smattheis** [comment]
@cpaelzer I would be happy to help out where I can. To do that, just a few questions:
1. With "confirm it works", do you refer to the PPA, i.e., the archive, or to open-vm-tools, i.e., the package? The PPA is perfect and for the package I'm still testing and will come back to that. (I encounter some lags with the 4K monitor support which was a heavy issue with older versions VMWare Tools, later no issue at all with the open-vm-tools 10.0.7 from standard Ubuntu PPAs, and now seems to have little issue again with the open-vm-tools 10.2.0, but maybe it's subjective so I will test that a bit more.) 
2. Which bug do you mean? The bug of vmhgfs-fuse in version 1.6.2.0 that resulted in delayed synchronization of shared folders and which was fixed in version 1.6.5.0 as of open-vm-tools version 10.1.0 and, therefore, still exists in open-vm-tools 10.0.7 as provided with standard PPAs? I can definitely confirm that, I just don't know how that works. Sorry. :)

**Comment by cpaelzer** [comment]
@smattheis 
For 1. - yes I refer to test the open-vm-tools made available via the PPA (Version 2:10.2.0-3ubuntu0.16.04.1~ppa1	at the moment)
(I already got a race in the new version reported for example).
Essentially anything that helps to make it better before making it available in xenial-backports is good to be known (even if it would be something we can't fix immediately, knowing about it is better).

For 2. - I meant an update chiming in on bug [1] that I linked above, this is more about showing that there is a wider need for backports of newer open-vm-tools on Xenial (or in general). I'm already on this particular bug, so your update there is not required - but in general it is nice to see that not only a single person wants something.

[1]: https://bugs.launchpad.net/ubuntu/+source/open-vm-tools/+bug/1741390

**Comment by smattheis** [comment]
@cpaelzer Okay, I got you. Thanks. 
1. Short question: Should this discussion be moved to launchpad? https://bugs.launchpad.net/ubuntu/+source/open-vm-tools/+bug/1741390 
2. I'm having  serious difficulties to install updates, i.e., `apt-get upgrade` fails and I'm not able to fix it with aptitude:
``` bash
$ sudo apt-get update                                       
Hit:1 http://ppa.launchpad.net/ci-train-ppa-service/3152/ubuntu xenial InRelease
Hit:2 http://ppa.launchpad.net/git-core/ppa/ubuntu xenial InRelease            
Hit:3 http://us.archive.ubuntu.com/ubuntu xenial InRelease                     
Hit:4 http://ppa.launchpad.net/webupd8team/java/ubuntu xenial InRelease        
Hit:5 http://security.ubuntu.com/ubuntu xenial-security InRelease              
Hit:6 http://us.archive.ubuntu.com/ubuntu xenial-updates InRelease       
Hit:7 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease
Reading package lists... Done
$ sudo apt-get upgrade
Reading package lists... Done
Building dependency tree       
Reading state information... Done
You might want to run 'apt-get -f install' to correct these.
The following packages have unmet dependencies:
 open-vm-tools-desktop : Depends: open-vm-tools (= 2:10.2.0-3ubuntu0.16.04.1~ppa1) but 2:10.2.0-3ubuntu0.16.04.1~ppa2 is installed
E: Unmet dependencies. Try using -f.
$ sudo apt-get -f install
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Correcting dependencies... Done
The following packages were automatically installed and are no longer required:
  linux-headers-4.4.0-109 linux-headers-4.4.0-109-generic linux-image-4.4.0-109-generic linux-image-extra-4.4.0-109-generic
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  open-vm-tools open-vm-tools-desktop
Suggested packages:
  cloud-init
The following NEW packages will be installed:
  open-vm-tools
The following packages will be upgraded:
  open-vm-tools-desktop
1 upgraded, 1 newly installed, 0 to remove and 15 not upgraded.
Need to get 0 B/667 kB of archives.
After this operation, 2,140 kB of additional disk space will be used.
Do you want to continue? [Y/n] Y
Selecting previously unselected package open-vm-tools-desktop.
(Reading database ... 440087 files and directories currently installed.)
Preparing to unpack .../open-vm-tools-desktop_2%3a10.2.0-3ubuntu0.16.04.1~ppa2_amd64.deb ...
Failed to stop run-vmblock-fuse.mount: Unit run-vmblock-fuse.mount not loaded.
dpkg: warning: subprocess old pre-removal script returned error exit status 5
dpkg: trying script from the new package instead ...
Failed to stop run-vmblock-fuse.mount: Unit run-vmblock-fuse.mount not loaded.
dpkg: error processing archive /var/cache/apt/archives/open-vm-tools-desktop_2%3a10.2.0-3ubuntu0.16.04.1~ppa2_amd64.deb (--unpack):
 subprocess new pre-removal script returned error exit status 5
/usr/bin/deb-systemd-helper: error: unable to read run-vmblock-fuse.mount
Failed to get unit file state for run-vmblock-fuse.mount: No such file or directory
run-vmblock-fuse.mount is a disabled or a static unit, not starting it.
Selecting previously unselected package open-vm-tools.
Preparing to unpack .../open-vm-tools_2%3a10.2.0-3ubuntu0.16.04.1~ppa2_amd64.deb ...
Unpacking open-vm-tools (2:10.2.0-3ubuntu0.16.04.1~ppa2) ...
Processing triggers for libc-bin (2.23-0ubuntu10) ...
Processing triggers for systemd (229-4ubuntu21.1) ...
Processing triggers for ureadahead (0.100.0-19) ...
Processing triggers for man-db (2.7.5-1) ...
Errors were encountered while processing:
 /var/cache/apt/archives/open-vm-tools-desktop_2%3a10.2.0-3ubuntu0.16.04.1~ppa2_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
$ sudo apt-get purge open-vm-tools open-vm-tools-desktop open-vm-tools-dkms
```
Now, open-vm-tools is uninstalled but open-vm-tools-desktop is installed. I cannot install, upgrade, or uninstall anything. Plus the VM is not working because open-vm-tools is uninstalled.

**Comment by cpaelzer** [comment]
@smattheis 
yes the discussion on the actual ppa test on https://bugs.launchpad.net/ubuntu/+source/open-vm-tools/+bug/1741390 please
The discussion on your install issue on an extra bug - I opened https://bugs.launchpad.net/ubuntu/+source/open-vm-tools/+bug/1751016 for you

**Comment by smattheis** [comment]
Thanks. Okay, I'm moving the discussion.

**Comment by markusd1984** [comment]
glad to hear somebody managed to fix it, thanks @smattheis 
but where can we locate the pre-removal script to edit?
